---
title: "Phyloseq PreProcessing"
author: "Calla Bush St George"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(echo = FALSE,
                      fig.align = "center",
                      fig.path = "../figures/02_PreProcessing/") # send any figure output to this folder 
```

# Goals

Here, we will process the data into a phyloseq object. 

- ASV table 
- Taxonomy Table 
- Track Reads (metadata)

Then, we will remove the following: 

1. Remove Chloroplasts
2. Remove Mitochondria. 
3. Removing ASVs from negative controls and also negative control samples.
4. Remove samples without "enough" reads. 

Finally, write data file of phyloseq output.

# Set-up

## Load libraries 
```{r load-libraries}
#devtools::install_github("joey711/phyloseq")
pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)
```


## Load Data 

### ASV Table
```{r load-asv-table}
# First, load asv table
load("combined_data/data/01_DADA2/ASV_counts.RData")

# Inspect asv_tab
head(asv_tab)[,20:25]

colnames(asv_tab) <-
  sub("_R1_filtered.fastq.gz$", "", colnames(asv_tab))


str(asv_tab)
head(asv_tab)[,20:25]
```

##  Taxonomy Table
```{r load-tax-table}
tax_df <- read.table("combined_data/data/01_DADA2/ASV_taxonomy.tsv",
                     sep = "\t", skip = 1)
head(tax_df)

# fix column names 
colnames(tax_df) <- c("asv_names", "Kingdom", "Phylum", "Class", "Order",
                      "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df)

# Taxonomy Table Matrix
tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "asv_names") %>%
  as.matrix()
```


##  Metadata
```{r load-metadata}
# Load in metadata
metadata_df <- read.csv("combined_data/data/combined_metadata.csv")
dim(metadata_df)
colnames(metadata_df)
head(metadata_df)

```


# Handoff to phyloseq
```{r phyloseq-handoff}

rownames(metadata_df) <-metadata_df$Run

# double check it's all good 
dim(asv_tab)
dim(tax_mat)
dim(metadata_df)

# Intuition check 
stopifnot(row.names(asv_tab) == row.names(tax_mat))

# Construct the phyloseq object 
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE),
                       sample_data(metadata_df),
                       tax_table(tax_mat))
raw_physeq

# Save this raw phyloseq object 
# Make new folder in data/ "02_Preprocessing"
save(raw_physeq, file = "combined_data/data/02_PreProcessing/raw_physeq.RData")
```

# Clean up the data

Remove: 

1. Chloroplasts
2. mitochondria  

```{r rm-mitos-chloros}
# Remind myself of tax table 
#View(tax_mat)

# Make new physeq without chloroplasts
noChloros_physeq <- 
  raw_physeq %>% 
  # rm chloroplasts
  subset_taxa(Order != "Chloroplast" | is.na(Order))
  
# How many taxa were chloroplasts? 
num_chloro_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloro_ASVs

# Intuition chek 
#noChloros_physeq %>%
#  tax_table() %>%
#  data.frame() %>%
#  View()

# remove mitochondria 
noChlorosMitos_physeq <- 
  noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

# How many mitochondrial ASVs? 
num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(noChlorosMitos_physeq)
num_mito_ASVs

noChlorosMitos_physeq

# How many total asvs were removed from chloros and mitos 
ntaxa(raw_physeq) - ntaxa(noChlorosMitos_physeq)
# proportion of asvs kept? 
ntaxa(noChlorosMitos_physeq)/ntaxa(raw_physeq)
```


# Evaulate and remove the control samples 

Take a look at the negative controls and then make a decision about whether or not to remove the ASVs that we found in our controls 

1. Negative controls 
2. ASVs found within the negative controls and their distribution in the samples. 
3. Evaluate the mock community 


## Negative Controls 
```{r neg-controls}
# create a vector of samples that were negative controls 
control_samples <- c("Neg1_S148", "Neg2_S249")

# Make a new phyloseq object of only the control samples 
control_physeq <- 
  noChlorosMitos_physeq %>%
  subset_samples(., Run %in% control_samples) %>%
  # force remove ASVs with 0 counts 
  prune_taxa(taxa_sums(.) > 0, .)


# vector of control asvs
control_ASVs <- 
  control_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV) %>%
  as.vector()

# Visualize raw abundance values of control ASVs
control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")

############# Check the control ASVs in the samples 
# Make new phyloseq object with only the control ASVs and ALL samples 
controlASV_physeq <- 
  noChlorosMitos_physeq %>%
  subset_taxa(., ASV %in% control_ASVs$ASV)

# Inspect it 
controlASV_physeq

# Let's take a look at the abundances 
controlASV_physeq %>%
  plot_bar(., "ASV", fill = "ASV", facet_grid = Sample_or_Control~.)

control_taxa <-
controlASV_physeq %>%
 tax_table() %>%
 data.frame()
view(control_taxa)

write.csv(df, 'data/control_taxa.csv')

# Would only like to keep ASV 1168, 1293, 1889, 2084, 3346, 3573, everything else is trash
controlASVs_toREMOVE <- 
  dplyr::setdiff(control_ASVs$ASV, c("ASV_1168", "ASV_1293", "ASV_1889",
                                     "ASV_2084", "ASV_3346", "ASV_3573"))
# double check 
length(controlASVs_toREMOVE)

### Make my new phloseq object 
noControlsChlorosMitos_physeq <- 
  noChlorosMitos_physeq %>%
  # Actually remove the control ASVs (except 1168, 1293, 1889, 2084, 3346, 3573)
  subset_taxa(., !(ASV %in% controlASVs_toREMOVE)) %>%
  # Remove control samples 
  subset_samples(., !(Run %in% control_samples))

# Inspect no controls object 
noControlsChlorosMitos_physeq

# After filtering, how many ASVs were removed? 
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noControlsChlorosMitos_physeq)
total_ASVs_rm
```


# Evaluate the Sequencing Depth 

```{r seq-depth, fig.width=12, fig.height=8}
# The current data object
noControlsChlorosMitos_physeq

# What is the library size/sequencing depth for each sample? 
seqSums_df <- 
  noControlsChlorosMitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "Run") %>%
  left_join(., metadata_df, by = "Run") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)
head(seqSums_df)

# Show the depth of samples 
seqSums_df %>%
  dplyr::select(Run, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

# plot it! 
seqSums_df %>%
  ggplot(aes(x=reorder(Run, TotalSeqs), y = TotalSeqs,
             fill = Run)) + 
  geom_bar(stat = "identity") 

# Density plot #Something is wrong with this
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = Host)) +
  geom_density(alpha = 0.5)

```

# Save Preprocessed Phyloseq Object
```{r save-physeq}

raw_preprocessed_physeq <- noControlsChlorosMitos_physeq

save(raw_preprocessed_physeq,
     file = "combined_data/data/02_PreProcessing/raw_preprocessed_physeq.RData")
```

# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```

