---
title: "Between sample/beta diversity of Microbes by gut section"
author: "Calla Bush St George"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/05_Community_Analysis/")
```

# Goals

1. Load in phyloseq data with rooted tree.
2. Evaluate sequencing depth and remove sample.
3. Normalize the read counts between samples.
4. Calculate community dissimilarities. Numbers between 0 and 1. If 0, completely similar versus if they are 1, then they are completely dissimilar. 
    a. *Sorensen*: Presence/ Absence. Weighted by number of shared taxa. Shared species as a binary-valye. Abundance-unweighted. 
    b. *Bray-Curtis*: Relative abundance. Weighted by number of shared taxa. Shared abundant species: abundance weighted. 
    c. *(Abundance) Weighted UNIFRAC*: Consider abundant species and where they fall on the tree. 
5. Visualize the community data with two unconstricted ordinations:
    a. *PCoA*: Linear method. Uses matrix algebra to calculate eigenvalye. Calculate how much variation is explained by each axis.  Can choose to view axis 1, 2, 3, etc. and plot them together. 
    b. *NMDS*: Non-linear method. Collapse multiple axes into two (or three) dimensions. Can see more axes of variation into fewer axes. Always need to report a stress value. (Ideally less than 0.15)
6. Run statistics with PERMANOVA and betadispR.

# Setup

## Load Libraries
```{r load-packages}
#install.packages("vegan")
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan,
               install = FALSE)
```

## Load in colors
```{r load-colors}
host_colors <- c(
  "Pomacanthus sexstriatus" = "navy",
  "Kyphosus sydneyanus" = "darkslategray2",
  "Kyphosus cinerascens" = "deeppink1",
  "Kyphosus vaigiensis" = "plum2",
  "Pomacanthus annularis" = "gold")
```


## Load in data 
```{r load-physeq}
# Load in rooted phylogenetic tree!
load("combined_data/data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
unrooted_physeq
midroot_physeq
```

# Explore Read Counts

## Raw Read Depth
```{r calculate-seq-depth}
# Sequence depth will inform us on how we want to normalize our data
# Calculate the total number of reads per sample
raw_total_seqs_df <-
  unrooted_physeq %>%
  # Calculate the sample read sums
  sample_sums() %>%
  data.frame()

# Name the column
colnames(raw_total_seqs_df)[1] <- "TotalSeqs"

head(raw_total_seqs_df)


# Make a histogram of raw reads
raw_seqs_histogram <-
  raw_total_seqs_df %>%
  ggplot(aes(x = TotalSeqs)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(limits = c(0, 50000)) +
  labs(title = "Raw Sequencing Depth Distribution") + 
  theme_bw()

```

## Remove lowly sequenced sample
```{r low-seq-sample}
raw_rooted_physeq <-
  midroot_physeq %>%
  # Remove lowly sequence sample that was outlier in alpha diversity analysis
  subset_samples(Run != "G196_III") %>%
  subset_samples(Run != "G197_III") %>%
  subset_samples(Run != "G197_IV") %>%
  subset_samples(Run != "G200_III") %>%
  subset_samples(Run != "G201_III") %>%
  subset_samples(Run != "SRR17060829") %>%
  subset_samples(Run != "SRR17060839") %>%
  subset_samples(Run != "SRR17060845") %>%
  subset_samples(Run != "G196_IV") %>%
  subset_samples(Run != "G199_III") %>%
  # Any ASVs unique to the sample will also be removed 
  prune_taxa(taxa_sums(.) > 0, .)

# Inspect
raw_rooted_physeq

# What is the minimum number of sequences
raw_rooted_physeq %>%
  sample_sums() %>%
  min()


```


## Normalize Read Counts
```{r scale-reads}

### scale_reads function and matround function
#################################################################################### 
# Function to scale reads: http://deneflab.github.io/MicrobeMiseq/ 
# Scales reads by 
# 1) taking proportions
# 2) multiplying by a given library size of n
# 3) rounding 
# Default for n is the minimum sample size in your library
# Default for round is floor

matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
  
  }

```

Rescale all reads so they all represent the count of the lowest number of sequence reads.
We will expect each sample to have # of reads around 270

This is where one might decide to use rarefaction to normalize the data. 

## Scale reads and check the distribution of the seq depth
```{r scale-physeq}
min(sample_sums(raw_rooted_physeq))

# Scale reads by the above function
scaled_rooted_physeq <-
  raw_rooted_physeq %>%
  scale_reads(round = "matround")

# Calculate read depth
## Look at total number of sequences in each sample and compare to what we had before

scaled_total_seqs_df <- 
  scaled_rooted_physeq %>%
  sample_sums() %>%
  data.frame()

head(scaled_total_seqs_df)

# Change first column name to be "TotalSeqs"
colnames(scaled_total_seqs_df)[1] <- "TotalSeqs"

# Inspect
head(scaled_total_seqs_df)

# Check range of data
min_seqs <-
  min(scaled_total_seqs_df)
max_seqs <-
 max(scaled_total_seqs_df)
# Range of seqs
max_seqs - min_seqs

# Plot histogram
scaled_total_seqs_df %>%
  ggplot(aes(x = TotalSeqs)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(limits = c(0, 300)) +
  labs(title = "Scaled Sequencing Depth at 270") + 
  theme_bw()

head(scaled_total_seqs_df)

```

# Calculate & Visualize Community Dissimiliarity

Exploratory analyses from the Paily & Shankar (2016) paper, which is using unconstrained ordination methods like PCoA. 

## Sorenson PCoA
```{r sorenson-pcoa}
# Calculate sorenson dissimularity: Abundance-unweighted of shared taxa
scaled_soren_pcoa <-  
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "bray", binary = TRUE)

#str(scaled_soren_pcoa)

# Plot the ordination 
soren_gut_section_pcoa <- plot_ordination(
  physeq = scaled_rooted_physeq,
  ordination = scaled_soren_pcoa,
  color = "Host",
  shape = "Host",
  title = "Sorensen PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = Host)) +
  scale_shape_manual(values = c(15,16,17,18,19)) +
  scale_color_manual(values = host_colors) + 
  theme_bw()
# Show the plot 
soren_gut_section_pcoa
```

Here, we are evaluating the shared taxa by presence/absence (abundance-unweighted) in the Sorensen metric. 

Note that:  

- Axis 1 = ~13% of variation  
- Axis 2 = ~12% of variation 

This means we explain 25% of the variation in the data in these two axes. 



## Bray-Curtis PCoA
```{r bray-PCoA}
# Calculate the BC distance
scaled_BC_pcoa <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "bray")

# Plot the PCoA
bray_gut_section_pcoa <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_BC_pcoa,
    color = "Host",
    shape = "Host",
    title = "Bray-Curtis PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = Host)) +
  scale_shape_manual(values = c(15,16,17,18,19)) +
  scale_color_manual(values = c(host_colors)) + 
  theme_bw()
bray_gut_section_pcoa
```

Here, we are evaluating the shared taxa and then weighting them by their abundances, which provides more influence for species that are more abundant. 

Note that:  

- Axis 1 = ~12% of variation  
- Axis 2 = ~12% of variation 

This means we explain 44% of the variation in the data in these two axes, which is almost exactly the same as the previous plot with the Sorensen Dissimilarity. 
Abundance does NOT seem to have an influence!!

## Weighted-Unifrac PCoA
```{r wUnifrac-PCoA}
# Calculate the BC distance
scaled_wUNI_pcoa <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "wunifrac")

# Plot the PCoA
wUNI_gut_section_pcoa <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_wUNI_pcoa,
    color = "Host",
    shape = "Host",
    title = "Weighted Unifrac PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = Host)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(host_colors)) + 
  theme_bw()
wUNI_gut_section_pcoa
```

Here, we are evaluating the shared taxa and then weighting them by their abundances, which provides more influence for species that are more abundant. 


Note that:   

- Axis 1 = ~45% of variation  
- Axis 2 = ~15% of variation 

This means we explain *60*% of the variation in the data in these two axes (!!!), which is significantly more than the previous plots with the taxonomic dissimilarity measures. 

## Combine PCoAs

Let's plot all three together into one plot to have a concise visualization of the three metrics. 
```{r pcoa-together, fig.width=8, fig.height=3.5}
(soren_gut_section_pcoa + theme(legend.position = "none")) + 
  (bray_gut_section_pcoa + theme(legend.position = "none")) + 
    (wUNI_gut_section_pcoa + theme(legend.position = "none"))
```

## NMDS 

### Weighted Unifrac

Since we did 3 of the dissimilarity metrics for the PCoA, let's just plot one example of them for the NMDS plotting. Here, we will use weighted Unifrac 
```{r wUnifrac-NMDS}
# Calculate the Weighted Unifrac distance
scaled_wUNI_nmds <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "NMDS",
    distance = "wunifrac")

# Plot the PCoA
wUNI_gut_section_nmds <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_wUNI_nmds,
    color = "Host",
    shape = "Host",
    title = "Weighted Unifrac NMDS") +
  geom_point(size=5, alpha = 0.5, aes(color = Host)) +
  scale_shape_manual(values = c(15,16,17,18,19)) +
  scale_color_manual(values = c(host_colors)) + 
  theme_bw()
wUNI_gut_section_nmds
```

We can see from above the plot that the stress value is ~0.121, which is close to being on the edge of acceptable.

In this case, I would always prefer to report the PCoA results because they are linear and provide a lot more post-hoc analyses to follow up with. In addition, it's helpful to only have 2 axes of variation and show how much variation is explained. 

```{r combined-wUnifrac, fig.width=6, fig.height=3.5}
(wUNI_gut_section_pcoa + theme(legend.position = "none")) + 
  (wUNI_gut_section_nmds + theme(legend.position = "none"))
```


# Statistical Significance Testing 

## PERMANOVA 
```{r PERMANOVA}
# Calculate all three of the distance matrices
scaled_sorensen_dist <- phyloseq::distance(scaled_rooted_physeq, method = "bray", binary = TRUE)
scaled_bray_dist <- phyloseq::distance(scaled_rooted_physeq, method = "bray")
scaled_wUnifrac_dist <- phyloseq::distance(scaled_rooted_physeq, method = "wunifrac")

# make a data frame from the sample_data
# All distance matrices will be the same metadata because they 
# originate from the same phyloseq object. 
metadata <- data.frame(sample_data(scaled_rooted_physeq))

# Adonis test
# In this example we are testing the hypothesis that the five stations
# that were collected have different centroids in the ordination space 
# for each of the dissimilarity metrics, we are using a discrete variable 
adonis2(scaled_sorensen_dist ~ Host, data = metadata)
adonis2(scaled_bray_dist ~ Host, data = metadata)
adonis2(scaled_wUnifrac_dist ~ Host, data = metadata)
```

Note that:  

- R2 = the percent variation explained. 
- F = the F-Statistic, which represents the importance value. 
- Pr(>F) = the pvalue 


Above, we see that the most variation is explained by the weighted unifrac, which explains ~41% of the variation in the data and also has the highest F-statistic.


## BetaDispR

The PERMANOVA is sensitive to variance/dispersion in the data. Therefore, we need to run a homogeneity of dispersion test to test for the sensitivity of our PERMANOVA results to variance. 
```{r betadispR}
# Homogeneity of Disperson test with beta dispr
# Sorensen 
beta_soren_gut_section <- betadisper(scaled_sorensen_dist, metadata$Host)
permutest(beta_soren_gut_section)

# Bray-curtis 
beta_bray_gut_section <- betadisper(scaled_bray_dist, metadata$Host)
permutest(beta_bray_gut_section)

# Weighted Unifrac 
beta_bray_gut_section <- betadisper(scaled_wUnifrac_dist, metadata$Host)
permutest(beta_bray_gut_section)
```

Above, our variance is impacted by host. 

# Session Information
 For Reproducibility
```{r session-info}
#Ensure reproducibility
devtools::session_info()
```
 


