---
title: "Combining ASV tables"
author: "Calla Bush St George"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "figures/01_DADA2/") #Send any figure output to this folder
```

# Goals for this document

1. Combine ASV sequence tables from Soh data, Alessandro data, Bikrim data, and Angert data

2. Assign taxonomy 

3. Create ASV count table

4. Save files for preprocessing

# Set up environment

## Load libraries
```{r load-libraries}
# Efficient package loading with pacman 
pacman::p_load(tidyverse, devtools, dada2, phyloseq, patchwork, DT, 
               install = FALSE)
```

## Set seed
```{r set-seed}
# Tries to ensure reproducibility when knitting document
set.seed(567890)
```


# Load in ASV sequence tables
```{r load-ASV-seq-tabs}
# Load in Angert ASV seq table
load("combined_data/data/angert_meta_analysis/noChimeras_ASV_table.RData")
noChimeras_ASV_table_angert <- noChimeras_ASV_table
rm(noChimeras_ASV_table)

#Trim 94 bp of right of sequence
colnames(noChimeras_ASV_table_angert) <- 
  substr(colnames(noChimeras_ASV_table_angert), 1, 142)

# Load Soh ASV table
load("combined_data/data/Soh_analysis/noChimeras_ASV_table.RData")
noChimeras_ASV_table_soh <- noChimeras_ASV_table
rm(noChimeras_ASV_table)

#Trim 8 bp of left of sequence
colnames(noChimeras_ASV_table_soh) <- 
  substr(colnames(noChimeras_ASV_table_soh), 8, 151)

# Load Bikrim ASV table
load("combined_data/data/Bikrim_analysis/noChimeras_ASV_table.RData")
noChimeras_ASV_table_bikrim <- noChimeras_ASV_table
rm(noChimeras_ASV_table)

#Trim 146 bp of left of sequence
colnames(noChimeras_ASV_table_bikrim) <- 
  substr(colnames(noChimeras_ASV_table_bikrim), 146, 289)

# Load Alessandro ASV tables
## Load 2011 Illumina run
load("combined_data/data/Alessandro_2011/noChimeras_ASV_table.RData")
noChimeras_ASV_table_alessandro_2011 <- noChimeras_ASV_table
rm(noChimeras_ASV_table)

#Trim 163 bp of left of sequence
colnames(noChimeras_ASV_table_alessandro_2011) <- 
  substr(colnames(noChimeras_ASV_table_alessandro_2011), 163, 305)

## Load 2017 Illumina run 
load("combined_data/data/Alessandro_2017/noChimeras_ASV_table.RData")
noChimeras_ASV_table_alessandro_2017 <- noChimeras_ASV_table
rm(noChimeras_ASV_table)

#Trim  bp of left of sequence
colnames(noChimeras_ASV_table_alessandro_2017) <- 
  substr(colnames(noChimeras_ASV_table_alessandro_2017), 185, 327)

```

# Combine ASV seq tables from different runs
```{r combine-ASV-seq-tabs}
noChimeras_ASV_table <-  
  mergeSequenceTables(
  table1 = noChimeras_ASV_table_angert, 
  table2 = noChimeras_ASV_table_soh, 
  table3 = noChimeras_ASV_table_bikrim, 
  table4 = noChimeras_ASV_table_alessandro_2011,
  table5 = noChimeras_ASV_table_alessandro_2017)
```

# Merge ASVs that are the same but different lengths
```{r merge-ASVs}
collapse_noChimeras_ASV_table <-
  collapseNoMismatch(seqtab = noChimeras_ASV_table)
```

# Assign taxonomy
Here we will use the Silva nr99 v138 database
```{r assign-taxonomy}
taxa_train <- 
  assignTaxonomy(collapse_noChimeras_ASV_table, 
                 "/workdir/in_class_data/taxonomy/silva_nr99_v138.1_train_set.fa.gz", 
                 multithread=6)

taxa_addSpecies <- 
  addSpecies(taxa_train, 
             "/workdir/in_class_data/taxonomy/silva_species_assignment_v138.1.fa.gz")

# Inspect the taxonomy 
taxa_print <- taxa_addSpecies # Removing sequence rownames for display only
rownames(taxa_print) <- NULL
# View(taxa_print)
```

# Prepare the data for export

## 1. ASV table 

Below, we will prepare the following: 

1. Two ASV Count tables: 
      a. With ASV seqs: ASV headers include the *entire* ASV sequence ~142bps.
      b. with ASV names: This includes re-written and shortened headers like ASV_1, ASV_2, etc, which will match the names in our fasta file below.  
2. `ASV_fastas`: A fasta file that we can use to build a tree for phylogenetic analyses (e.g. phylogenetic alpha diversity metrics or UNIFRAC dissimilarty).  

### Finalize ASV Count Tables 
```{r prepare-ASVcount-table}
########### COUNT TABLE ###############
############## Modify the ASV names and then save a fasta file!  ############## 
# Give headers more manageable names
# First pull the ASV sequences
asv_seqs <- colnames(collapse_noChimeras_ASV_table)
asv_seqs[1:5]

# make headers for our ASV seq fasta file, which will be our asv names
asv_headers <- vector(dim(collapse_noChimeras_ASV_table)[2], mode = "character")
asv_headers[1:5]

# loop through vector and fill it in with ASV names 
for (i in 1:dim(collapse_noChimeras_ASV_table)[2]) {
  asv_headers[i] <- paste(">ASV", i, sep = "_")
  }

# intitution check
asv_headers[1:5]

##### Rename ASVs in table then write out our ASV fasta file! 
#View(collapse_noChimeras_ASV_table)
asv_tab <- t(collapse_noChimeras_ASV_table)
#View(asv_tab)

## Rename our asvs! 
row.names(asv_tab) <- sub(">", "", asv_headers)
#View(asv_tab)
```


## 2. Taxonomy Table 
```{r prepare-tax-table}
# Inspect the taxonomy table
#View(taxa_addSpecies)

##### Prepare tax table 
# Add the ASV sequences from the rownames to a column 
new_tax_tab <- 
  taxa_addSpecies%>%
  as.data.frame() %>%
  rownames_to_column(var = "ASVseqs") 
head(new_tax_tab)

# intution check 
stopifnot(new_tax_tab$ASVseqs == colnames(collapse_noChimeras_ASV_table))

# Now let's add the ASV names 
rownames(new_tax_tab) <- rownames(asv_tab)
head(new_tax_tab)

### Final prep of tax table. Add new column with ASV names 
asv_tax <- 
  new_tax_tab %>%
  # add rownames from count table for phyloseq handoff
  mutate(ASV = rownames(asv_tab)) %>%
  # Resort the columns with select
  dplyr::select(Kingdom, Phylum, Class, Order, Family, Genus, Species, ASV, ASVseqs)

head(asv_tax)

# Intuition check
stopifnot(asv_tax$ASV == rownames(asv_tax), rownames(asv_tax) == rownames(asv_tab))
```



# Write `01_DADA2` files

Now, we will write the files! We will write the following to the `data/01_DADA2/` folder. We will save both as files that could be submitted as supplements AND as .RData objects for easy loading into the next steps into R.:  

1. `ASV_counts.tsv`: ASV count table that has ASV names that are re-written and shortened headers like ASV_1, ASV_2, etc, which will match the names in our fasta file below. This will also be saved as `data/01_DADA2/ASV_counts.RData`.
2. `ASV_counts_withSeqNames.tsv`: This is generated with the data object in this file known as `noChimeras_ASV_table`. ASV headers include the *entire* ASV sequence ~250bps.  In addition, we will save this as a .RData object as `data/01_DADA2/noChimeras_ASV_table.RData` as we will use this data in `analysis/02_PreProcessing.Rmd` to assign the taxonomy from the sequence headers.  
3. `ASVs.fasta`: A fasta file output of the ASV names from `ASV_counts.tsv` and the sequences from the ASVs in `ASV_counts_withSeqNames.tsv`. A fasta file that we can use to build a tree for phylogenetic analyses (e.g. phylogenetic alpha diversity metrics or UNIFRAC dissimilarty).  
4. We will also make a copy of `ASVs.fasta` in `data/02_PreProcessing/` to be used for the taxonomy classification in the next step in the workflow.  
5. Write out the taxonomy table
6. `track_read_counts.RData`: To track how many reads we lost throughout our workflow that could be used and plotted later. We will add this to the metadata in `analysis/02_PreProcessing.Rmd`.   


```{r save-files}
# FIRST, we will save our output as regular files, which will be useful later on. 
# Save to regular .tsv file 
# Write BOTH the modified and unmodified ASV tables to a file!
# Write count table with ASV numbered names (e.g. ASV_1, ASV_2, etc)
write.table(asv_tab, "combined_data/data/01_DADA2/ASV_counts.tsv", sep = "\t", quote = FALSE, col.names = NA)
# Write count table with ASV sequence names
write.table(noChimeras_ASV_table, "combined_data/data/01_DADA2/ASV_counts_withSeqNames.tsv",
            sep = "\t", quote = FALSE, col.names = NA)
# Write out the fasta file for reference later on for what seq matches what ASV
asv_fasta <- c(rbind(asv_headers, asv_seqs))
# Save to a file!
write(asv_fasta, "combined_data/data/01_DADA2/ASVs.fasta")


# SECOND, let's save the taxonomy tables 
# Write the table 
write.table(asv_tax, "combined_data/data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", quote = FALSE, col.names = NA)


# THIRD, let's save to a RData object 
# Each of these files will be used in the analysis/02_PreProcessing
# RData objects are for easy loading :) 
save(collapse_noChimeras_ASV_table,
     file = "combined_data/data/01_DADA2/noChimeras_ASV_table.RData")
save(asv_tab, file = "combined_data/data/01_DADA2/ASV_counts.RData")
# Did not track counts
```


##Session information
```{r session-info}
#Ensure reproducibility
devtools::session_info()
```




